packages/patrol_cli/lib/src/coverage/coverage_tool.dart

```diff
@@ -0,0 +1,213 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:adb/adb.dart';
+import 'package:coverage/coverage.dart' as coverage;
+import 'package:dispose_scope/dispose_scope.dart';
+import 'package:file/file.dart';
+import 'package:glob/glob.dart';
+import 'package:patrol_cli/src/base/logger.dart';
+import 'package:patrol_cli/src/coverage/device_to_host_port_transformer.dart';
+import 'package:patrol_cli/src/coverage/vm_connection_details.dart';
+import 'package:patrol_cli/src/devices.dart';
+import 'package:process/process.dart';
+import 'package:rxdart/rxdart.dart';
+import 'package:vm_service/vm_service_io.dart';
+
+class CoverageTool {
+  CoverageTool({
+    required FileSystem fs,
+    required ProcessManager processManager,
+    required Adb adb,
+    required DisposeScope parentDisposeScope,
+  })  : _fs = fs,
+        _processManager = processManager,
+        _adb = adb,
+        _disposeScope = DisposeScope() {
+    _disposeScope.disposedBy(parentDisposeScope);
+  }
+
+  final FileSystem _fs;
+  final ProcessManager _processManager;
+  final Adb _adb;
+  final DisposeScope _disposeScope;
+
+  Future<void> run({
+    required String flutterPackageName,
+    required TargetPlatform platform,
+    required Logger logger,
+    required Set<Glob> ignoreGlobs,
+  }) async {
+    final homeDirectory =
+        Platform.environment['HOME'] ?? Platform.environment['USERPROFILE'];
+    final hitMap = <String, coverage.HitMap>{};
+
+    await _disposeScope.run(
+      (scope) async {
+        final logsProcess = await _processManager.start(
+          ['flutter', 'logs'],
+          workingDirectory: homeDirectory,
+          runInShell: true,
+        )
+          ..disposedBy(scope);
+
+        final vmConnectionDetailsStream = logsProcess.stdout
+            .transform(utf8.decoder)
+            .transform(const LineSplitter())
+            .map(VMConnectionDetails.tryExtractFromLogs)
+            .whereNotNull()
+            .transform(
+              DeviceToHostPortTransformer(
+                processManager: _processManager,
+                devicePlatform: platform,
+                adb: _adb,
+                logger: logger,
+              ),
+            )
+            .whereNotNull()
+            .asBroadcastStream();
```

Comment:
If `rxdart` is used only for `whereNotNull` here, I'd consider using plain old `map` instead. Let's keep dependencies to reasonable minimum. No strong opinion though.